File: core/src/com/unciv/logic/battle/i_combatant.py
from abc import ABC, abstractmethod
from typing import Optional, cast
from com.unciv.logic.civilization import Civilization
from com.unciv.logic.map.tile import Tile
from com.unciv.models import UncivSound
from com.unciv.models.ruleset.unit import UnitType
from com.unciv.logic.battle.city_combatant import CityCombatant
from com.unciv.logic.battle.map_unit_combatant import MapUnitCombatant

class ICombatant(ABC):
    @abstractmethod
    def get_name(self) -> str:
        pass

    @abstractmethod
    def get_health(self) -> int:
        pass

    @abstractmethod
    def get_max_health(self) -> int:
        pass

    @abstractmethod
    def get_unit_type(self) -> UnitType:
        pass

    @abstractmethod
    def get_attacking_strength(self) -> int:
        pass

    @abstractmethod
    def get_defending_strength(self, attacked_by_ranged: bool = False) -> int:
        pass

    @abstractmethod
    def take_damage(self, damage: int) -> None:
        pass

    @abstractmethod
    def is_defeated(self) -> bool:
        pass

    @abstractmethod
    def get_civ_info(self) -> Civilization:
        pass

    @abstractmethod
    def get_tile(self) -> Tile:
        pass

    @abstractmethod
    def is_invisible(self, to: Civilization) -> bool:
        pass

    @abstractmethod
    def can_attack(self) -> bool:
        pass

    @abstractmethod
    def matches_filter(self, filter: str, multi_filter: bool = True) -> bool:
        """Implements UniqueParameterType.CombatantFilter"""
        pass

    @abstractmethod
    def get_attack_sound(self) -> UncivSound:
        pass

    def is_melee(self) -> bool:
        return not self.is_ranged()

    def is_ranged(self) -> bool:
        if isinstance(self, CityCombatant):
            return True
        unit_combatant = cast(MapUnitCombatant, self)
        return unit_combatant.unit.base_unit.is_ranged()

    def is_air_unit(self) -> bool:
        if isinstance(self, CityCombatant):
            return False
        unit_combatant = cast(MapUnitCombatant, self)
        return unit_combatant.unit.base_unit.is_air_unit()

    def is_water_unit(self) -> bool:
        if isinstance(self, CityCombatant):
            return False
        unit_combatant = cast(MapUnitCombatant, self)
        return unit_combatant.unit.base_unit.is_water_unit

    def is_land_unit(self) -> bool:
        if isinstance(self, CityCombatant):
            return False
        unit_combatant = cast(MapUnitCombatant, self)
        return unit_combatant.unit.base_unit.is_land_unit

    def is_city(self) -> bool:
        return isinstance(self, CityCombatant)

    def is_civilian(self) -> bool:
        if not isinstance(self, MapUnitCombatant):
            return False
        return self.unit.is_civilian()

File: rust_port/src/battle/i_combatant.rs
use crate::civilization::Civilization;
use crate::map::tile::Tile;
use crate::models::unciv_sound::UncivSound;
use crate::models::unit_type::UnitType;
use std::sync::Arc;

/// Trait for entities that can participate in combat
pub trait ICombatant: Clone + std::fmt::Debug {
    /// Get the name of the combatant
    fn get_name(&self) -> String;

    /// Get the current health of the combatant
    fn get_health(&self) -> i32;

    /// Get the maximum health of the combatant
    fn get_max_health(&self) -> i32;

    /// Get the unit type of the combatant
    fn get_unit_type(&self) -> UnitType;

    /// Get the attacking strength of the combatant
    fn get_attacking_strength(&self) -> i32;

    /// Get the defending strength of the combatant
    fn get_defending_strength(&self, attacked_by_ranged: bool) -> i32;

    /// Apply damage to the combatant
    fn take_damage(&mut self, damage: i32);

    /// Check if the combatant is defeated
    fn is_defeated(&self) -> bool;

    /// Get the civilization information of the combatant
    fn get_civ_info(&self) -> Arc<Civilization>;

    /// Get the tile the combatant is on
    fn get_tile(&self) -> Arc<Tile>;

    /// Check if the combatant is invisible to a specific civilization
    fn is_invisible(&self, to: &Civilization) -> bool;

    /// Check if the combatant can attack
    fn can_attack(&self) -> bool;

    /// Check if the combatant matches a filter
    ///
    /// Implements UniqueParameterType.CombatantFilter
    fn matches_filter(&self, filter: &str, multi_filter: bool) -> bool;

    /// Get the attack sound of the combatant
    fn get_attack_sound(&self) -> UncivSound;

    /// Check if the combatant is a melee unit
    fn is_melee(&self) -> bool {
        !self.is_ranged()
    }

    /// Check if the combatant is a ranged unit
    fn is_ranged(&self) -> bool;

    /// Check if the combatant is an air unit
    fn is_air_unit(&self) -> bool;

    /// Check if the combatant is a water unit
    fn is_water_unit(&self) -> bool;

    /// Check if the combatant is a land unit
    fn is_land_unit(&self) -> bool;

    /// Check if the combatant is a city
    fn is_city(&self) -> bool;

    /// Check if the combatant is a civilian unit
    fn is_civilian(&self) -> bool;
}